cep-repodata-next.md

```diff
@@ -0,0 +1,119 @@
+| Implementation | Flags: https://github.com/conda/rattler/pull/1040, Optional: https://github.com/conda/rattler/pull/1019, Conditional: https://github.com/prefix-dev/resolvo/pull/101  |
+
+
+The conda ecosystem finds and resolves packages based on "repodata" - the main index metadata for all artifacts in the condaverse.
+
+Repodata has been a stable format for a long time now. It generally consists of at least the following fields:
+
+```yaml
+name: name of the package
+version: version of the package
+build_string: build string of the package
+build_number: build number of the package
+depends: [MatchSpec] dependencies of the package, expressed as "triplet" matchspec
+constrains: [MatchSpec] constraints that the package adds to the resolution aka optional dependencies
+```
+
+All other fields are mostly for metadata purposes and not listed.
+
+With this CEP we would like to add 3 new fields to a proposed "repodata.v2" format.
+
+The fields serve three different purposes:
+
+- `extras`: optional dependency sets as known from the PyPI world. For examples, `sqlalchemy` might be a small base package that defines a number of extras such as `mysql`, `postgres`, `sqlite` that would pull in dependency sets as needed
+- `conditional` dependencies, also widely known from the Python world. These are activated only when the condition is true. For example, certain dependencies such as `pywin32` are only relevant on Windows and not on macOS or Linux.
+- `flags` are used to make it easier to select variants. Compiled packages can often be compiled with different options which results in different variants (for example, Debug vs. Release builds). With `flags` it will be trivial to select the preferred build with a syntax such as `foobar[flags=['release']]`. Flags are free-form and can be used by distributions such as conda-forge to differentiate between gpu and non-gpu builds as well.
+
+## Extra dependency sets
+
+We want to define a new `extras` key in `RepoData`. The key will be a dictionary mapping from String to list of MatchSpecs:
+
+```yaml
+name: sqlalchemy
+version: 1.0.0
+depends:
+ - python >=3.8
+extras:
+  sqlite:
+    - sqlite >=1.5
+    - py-sqlite-adapter 1.0
+  postgres:
+    - postgres >=3.5
+    - pyxpgres >=8
+```
+
+When a user, or a dependency, selects an extra through a MatchSpec, the extra and it's dependencies are "activated". This is conceptually the same as having three packages with "exact" dependencies from the "extra" to the base package: `sqlalchemy`, `sqlalchemy-sqlite` and `sqlalchemy-postgres` – which is the workaround currently employed by a number of packages on conda-forge.
+
+## Conditional dependencies
+
+Conditional dependencies are activated when the condition is true. The most straight-forward conditions are `__unix`, `__win` and other platform specifiers. However, we would also like to support matchspecs in conditions such as `python >=3`.
```

Comment:
I very much like this idea. An implementation note: while `if __unix` is reasonably easy to implement because it is "static", `if python <3.8` is conceptually much harder as it is not something that can be decided ahead of solving. I requires to be able to adapt the dependencies of a package as partial candidates are investigated during solve.

Reply:
This is the PR that implements this for resolvo: https://github.com/prefix-dev/resolvo/pull/136